File: swarm/app/tools/kakshya.py
"""Kakshya micro-compartments — daily precision via Ashtakavarga bindus.

Each 30° sign is divided into 8 Kakshyas of 3°45' each, ruled by planets
in order: Saturn, Jupiter, Mars, Sun, Venus, Mercury, Moon, Ascendant.

A transiting planet yields good results on days when it traverses a Kakshya
whose lord contributed a bindu in that planet's BAV for that sign.
"""

from __future__ import annotations

KAKSHYA_LORDS = ["Saturn", "Jupiter", "Mars", "Sun", "Venus", "Mercury", "Moon", "Asc"]
KAKSHYA_SPAN = 30.0 / 8  # 3.75°


def get_kakshya(longitude: float) -> dict:
    """Determine which Kakshya a planet occupies within its sign.

    Args:
        longitude: Sidereal longitude (0-360°).

    Returns:
        {kakshya_index (0-7), kakshya_lord, degree_in_sign, sign_index}
    """
    sign_index = int(longitude // 30) % 12
    degree_in_sign = longitude % 30
    kakshya_index = min(int(degree_in_sign / KAKSHYA_SPAN), 7)
    return {
        "kakshya_index": kakshya_index,
        "kakshya_lord": KAKSHYA_LORDS[kakshya_index],
        "degree_in_sign": round(degree_in_sign, 4),
        "sign_index": sign_index,
    }


def check_kakshya_bindu(
    transiting_planet: str,
    transit_longitude: float,
    bhinna_av: list[dict],
) -> dict:
    """Check if the Kakshya lord contributed a bindu for this transit.

    Args:
        transiting_planet: Name of the transiting planet.
        transit_longitude: Current sidereal longitude of the planet.
        bhinna_av: List of BAV entries [{planet, bindus[12], total}].

    Returns:
        {planet, sign_index, kakshya_lord, kakshya_index, has_bindu, bindu_count}
    """
    kak = get_kakshya(transit_longitude)
    sign_idx = kak["sign_index"]
    kakshya_lord = kak["kakshya_lord"]

    planet_bav = next((b for b in bhinna_av if b["planet"] == transiting_planet), None)
    if not planet_bav:
        return {
            "planet": transiting_planet,
            "sign_index": sign_idx,
            "kakshya_lord": kakshya_lord,
            "kakshya_index": kak["kakshya_index"],
            "has_bindu": False,
            "bindu_count": 0,
        }

    bindu_count = planet_bav["bindus"][sign_idx]

    has_bindu = bindu_count > 0

    return {
        "planet": transiting_planet,
        "sign_index": sign_idx,
        "kakshya_lord": kakshya_lord,
        "kakshya_index": kak["kakshya_index"],
        "has_bindu": has_bindu,
        "bindu_count": bindu_count,
    }


def daily_kakshya_score(
    transit_longitudes: dict[str, float],
    bhinna_av: list[dict],
) -> dict:
    """Compute the daily Kakshya score across all transiting planets.

    A score of 4+ active bindus = highly successful day.
    A score of 0-2 = day of struggle.
    """
    results = []
    active_count = 0
    for planet, lon in transit_longitudes.items():
        check = check_kakshya_bindu(planet, lon, bhinna_av)
        results.append(check)
        if check["has_bindu"]:
            active_count += 1

    if active_count >= 5:
        quality = "excellent"
    elif active_count >= 4:
        quality = "good"
    elif active_count >= 3:
        quality = "average"
    else:
        quality = "challenging"

    return {
        "active_bindus": active_count,
        "quality": quality,
        "details": results,
    }

File: src/lib/astro-tools/kakshya.ts
// =============================================================================
// Kakshya (Sub-division) & Ashtakavarga Bindu Transit Analysis
// =============================================================================
// Each sign (30 degrees) is divided into 8 kakshyas of 3.75 degrees each,
// ruled by Saturn, Jupiter, Mars, Sun, Venus, Mercury, Moon, Ascendant
// (in that fixed order). When a planet transits a kakshya whose lord has
// contributed a bindu in Bhinna Ashtakavarga, the transit is favorable.
// =============================================================================

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface KakshyaInfo {
  /** 0-based kakshya index within the sign (0-7) */
  kakshyaIndex: number;
  /** Lord of this kakshya */
  kakshyaLord: string;
  /** Degree position within the sign (0-30) */
  degreeInSign: number;
  /** 0-based sign index (0-11) */
  signIndex: number;
}

export interface KakshyaBinduResult {
  /** The transiting planet */
  planet: string;
  /** Kakshya information at the transit longitude */
  kakshya: KakshyaInfo;
  /** Number of bindus the planet has in this sign (from Bhinna AV) */
  bindusInSign: number;
  /** Whether the kakshya lord contributed a bindu (favorable sub-transit) */
  kakshyaLordHasBindu: boolean;
  /** Quality label — binary: the compartment lord either gave a bindu or didn't. */
  quality: 'favorable' | 'unfavorable';
}

export interface DailyKakshyaDetail {
  planet: string;
  kakshya: KakshyaInfo;
  binduActive: boolean;
}

export interface DailyKakshyaScore {
  /** Count of planets whose kakshya lord has a bindu */
  activeBindus: number;
  /** Overall quality: 'good' if majority active, 'poor' otherwise */
  quality: 'good' | 'average' | 'poor';
  /** Per-planet details */
  details: DailyKakshyaDetail[];
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/**
 * The 8 kakshya lords in fixed order within every sign.
 * Kakshya 0 (0-3.75 deg) = Saturn, Kakshya 1 (3.75-7.5 deg) = Jupiter, etc.
 */
export const KAKSHYA_LORDS: string[] = [
  'Saturn',
  'Jupiter',
  'Mars',
  'Sun',
  'Venus',
  'Mercury',
  'Moon',
  'Asc',
];

/** Each kakshya spans 3.75 degrees (30 / 8). */
export const KAKSHYA_SPAN = 3.75;

// ---------------------------------------------------------------------------
// Core Functions
// ---------------------------------------------------------------------------

/**
 * Determine which kakshya a given longitude falls in.
 *
 * @param longitude - Sidereal longitude (0-360)
 * @returns KakshyaInfo with sign index, degree in sign, kakshya index, and lord
 */
export function getKakshya(longitude: number): KakshyaInfo {
  // Normalize longitude to 0-360
  let normLong = longitude % 360;
  if (normLong < 0) normLong += 360;

  const signIndex = Math.floor(normLong / 30);
  const degreeInSign = normLong - signIndex * 30;
  const kakshyaIndex = Math.min(Math.floor(degreeInSign / KAKSHYA_SPAN), 7);
  const kakshyaLord = KAKSHYA_LORDS[kakshyaIndex];

  return { kakshyaIndex, kakshyaLord: kakshyaLord ?? 'Asc', degreeInSign, signIndex };
}

/**
 * Per-contributor Bhinnashtakavarga attribution — see
 * astro-engine/calculations/ashtakavarga.ts's calculateBhinnaAshtakavargaDetailed,
 * which produces this shape. contributions[contributorName][signIndex] is 1
 * if that contributor gave a bindu to `planet` in that sign, else 0.
 */
export interface DetailedBhinnaAshtakavarga {
  planet: string;
  contributions: Record<string, number[]>;
}

/**
 * Check whether a planet's CURRENT kakshya (3°45' compartment) lord actually
 * contributed a bindu to that planet in the sign it's transiting — the real
 * classical rule, not a shortcut. A transiting planet doesn't benefit from a
 * sign's whole bindu total the instant it enters the sign; it first crosses
 * Saturn's kakshya, then Jupiter's, and so on, and only gets a favorable
 * sub-window while it sits in a compartment whose lord specifically gave it
 * a bindu there. This is why the result changes as the planet moves through
 * the sign, several times over the sign's 30 degrees — unlike a whole-sign
 * total, which is constant for as long as the planet stays in that sign.
 *
 * @param transitingPlanet - Name of the planet transiting
 * @param transitLongitude - Current sidereal longitude of the planet
 * @param detailedBav - Per-contributor Bhinnashtakavarga detail (see
 *   calculateBhinnaAshtakavargaDetailed), NOT the collapsed per-sign totals.
 * @returns KakshyaBinduResult
 */
export function checkKakshyaBindu(
  transitingPlanet: string,
  transitLongitude: number,
  detailedBav: DetailedBhinnaAshtakavarga[],
): KakshyaBinduResult {
  const kakshya = getKakshya(transitLongitude);

  const planetAv = detailedBav.find((b) => b.planet === transitingPlanet);
  const contributions = planetAv?.contributions;

  // Whole-sign total, kept for context/display — no longer what decides
  // favorability, but still useful to report alongside the compartment verdict.
  let bindusInSign = 0;
  if (contributions) {
    for (const contributorBindus of Object.values(contributions)) {
      bindusInSign += contributorBindus[kakshya.signIndex] ?? 0;
    }
  }

  // The actual rule: does THIS kakshya's specific lord contribute a bindu to
  // this planet in this sign? Independent of the sign's total bindu count.
  const kakshyaLordHasBindu =
    (contributions?.[kakshya.kakshyaLord]?.[kakshya.signIndex] ?? 0) === 1;

  const quality: 'favorable' | 'unfavorable' = kakshyaLordHasBindu ? 'favorable' : 'unfavorable';

  return {
    planet: transitingPlanet,
    kakshya,
    bindusInSign,
    kakshyaLordHasBindu,
    quality,
  };
}

/**
 * Compute an aggregate daily kakshya score across multiple transiting planets.
 *
 * @param transitLongitudes - Record mapping planet names to their sidereal longitudes
 * @param detailedBav - Per-contributor Bhinnashtakavarga detail (see
 *   calculateBhinnaAshtakavargaDetailed), NOT the collapsed per-sign totals.
 * @returns DailyKakshyaScore with per-planet details and overall quality
 */
export function dailyKakshyaScore(
  transitLongitudes: Record<string, number>,
  detailedBav: DetailedBhinnaAshtakavarga[],
): DailyKakshyaScore {
  const details: DailyKakshyaDetail[] = [];
  let activeCount = 0;

  for (const [planet, longitude] of Object.entries(transitLongitudes)) {
    const result = checkKakshyaBindu(planet, longitude, detailedBav);
    const binduActive = result.kakshyaLordHasBindu;
    if (binduActive) activeCount++;

    details.push({
      planet,
      kakshya: result.kakshya,
      binduActive,
    });
  }

  const total = details.length;
  let quality: 'good' | 'average' | 'poor';
  if (total === 0) {
    quality = 'average';
  } else if (activeCount / total >= 0.6) {
    quality = 'good';
  } else if (activeCount / total >= 0.4) {
    quality = 'average';
  } else {
    quality = 'poor';
  }

  return {
    activeBindus: activeCount,
    quality,
    details,
  };
}
